sklearn/utils/validation.py

```diff
@@ -1259,3 +1270,42 @@ def inner_f(*args, **kwargs):
         kwargs.update({k: arg for k, arg in zip(all_args, args)})
         return f(**kwargs)
     return inner_f
+
+
+def _check_fit_params(X, fit_params, indices=None):
+    """Check and validate the parameters passed during `fit`.
+
+    Parameters
+    ----------
+    X : array-like of shape (n_samples, n_features)
+        Data array.
+
+    fit_params : dict
+        Dictionary containing the parameters passed at fit.
+
+    indices : array-like of shape (n_samples,), default=None
+        Indices to be selected if the parameter has the same size than `X`.
+
+    Returns
+    -------
+    fit_params_validated : dict
+        Validated parameters. We ensure that the values support indexing.
+    """
+    from . import _safe_indexing
+    fit_params_validated = {}
+    for param_key, param_value in fit_params.items():
+        if (not _is_arraylike(param_value) or
+                _num_samples(param_value) != _num_samples(X)):
```

Comment:
so we also support non-sample-aligned fit params? I though we wanted to support only the scalars for now.

Also, shouldn't we pass all non-scalars to `_make_indexable`?

Reply:
> so we also support non-sample-aligned fit params

At least it seems we were supporting it. This is the code used within the cross-validation originally.

Reply:
> so we also support non-sample-aligned fit params? I though we wanted to support only the scalars for now.

No, we want to be backward compatible, which means supporting non-sample-aligned fit params

Reply:
Fair, but we should still make sure the non-sample-aligned params are also not implementing `__array_function__`, I think.

Reply:
Let the downstream estimator deal with it for unaligned parameters I reckon. Pass them untouched

Reply:
For this specific PR I think we should just restore compatibility with the prior behavior.

For the future API, I am not so sure but I think I would also be in favor of passing the params untouched?